### PR TITLE
Add `LineRange` type

### PR DESCRIPTION
**Pull request by lionel-**
@DavisVaughan I didn't like using `TextRange` for lines in the end. `TextSize` tries hard to make the offset an opaque type and I needed lots of `.into()`. So I created our own type around an `std::ops::Range`. It mirrors (partially) the API of `TextRange` so that we don't need to think about how to use the range as much. If we really need genericity over text ranges and line ranges at some point we can create a trait. WDYT?

The type is implemented in a new `ark/coordinates` module, which should become a crate if also needed in the LSP (which is not the case at the moment in principle - we'll make it an LSP request but ideally it's a jupyter request).

### CODE REVIEW

**Review by DavisVaughan — APPROVED** [review]

**Comment by DavisVaughan** [comment]
I think `TextSize` making the `u32` opaque is nice if you use this data structure everywhere from the ground up

File: crates/ark/src/coordinates/line_range.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    pub fn contains(&self, line: u32) -> bool {
+        self.inner.contains(&line)
+    }
+
+    pub fn cover(&self, other: Self) -> Self {
+        let start = cmp::min(self.start(), other.start());
+        let end = cmp::max(self.end(), other.end());
+        Self::new(start, end)
+    }
```
[the remark this answers is not in the document]
Comment (lionel-) [reply crates/ark/src/coordinates/line_range.rs:32]: hmm... I'm not familiar with "union of ranges". What you say about covering the gap makes it a bit strange? In any case I think we should follow `text_size::TextRange`'s API so it's uniform across range types.

**Comment by lionel-** [comment]
> I think TextSize making the u32 opaque is nice if you use this data structure everywhere from the ground up

I can see how it makes sense with byte offsets where lots of assumptions are baked in, such as line endings, encodings, etc. That makes it hard to compare with offsets coming from external sources. However in the case of lines, which is a much more straightforward property of a text, it's practical to compare them to raw numbers coming from a loop or some external source.

File: crates/ark/src/coordinates/line_range.rs [hunk 2]
```diff
@@ -0,0 +1,39 @@
+use std::cmp;
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct LineRange {
+    pub inner: std::ops::Range<u32>,
```
[the remark this answers is not in the document]
Comment (lionel-) [reply crates/ark/src/coordinates/line_range.rs:5]: The accessor functions are part of the `text_size::TextRange` API where `start` and `end` are private.

[the remark this answers is not in the document]
Comment (lionel-) [reply crates/ark/src/coordinates/line_range.rs:5]: I removed the dependency on `std::ops::Range` but didn't change the API to keep it in line with `text_size`. But I'm happy to discuss this again and review the approach if needed.

**Pull request merged**